syntax = "proto3";

package raft;

service RaftConsensus {
  rpc RequestVote(VoteRequest) returns (VoteReply);
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesReply);
}

service RaftClient {
  rpc ClientRequest(ClientCommand) returns (ClientReply);
}

message VoteRequest {
  string candidate_id = 1;
  int32 term = 2;
  int32 last_log_index = 3;
  int32 last_log_term = 4;
}

message VoteReply {
  string voter_id = 1;
  int32 term = 2;
  bool vote_granted = 3;
}

message LogEntry {
  int32 term = 1;
  int32 index = 2;
  string command = 3;
}

message AppendEntriesRequest {
  string leader_id = 1;
  int32 term = 2;
  repeated LogEntry entries = 3;
  int32 commit_index = 4;
}

message AppendEntriesReply {
  string follower_id = 1;
  int32 term = 2;
  bool success = 3;
}

message ClientCommand {
  string command = 1;
}

message ClientReply {
  bool success = 1;
  string leader_id = 2;
  int32 commit_index = 3;
  string message = 4;
}
